### PR TITLE
First attempt at wrapping ABCpy - work-in-progress

### PR DESCRIPTION
**Pull request by LoryPack**
Hello, 

I thought I'd open this pull request to keep track of my progress and ask for feedback (even if I don't think this is ready to merge yet). 

I've done my first attempt at wrapping ABCpy. For now I've wrapped the Rejection ABC algorithm (I know you've that already, but that was the easiest one to do). 

It seems to work (I've added a `try_ABCpy.py` file if you want to experiment with it), but I have made quite some simplifications as I did not know how to do some things. Precisely:
- It looks like you have two separate parameters: `num_simulations` which is the total number of simulations which you allow, and then `num_samples` which is the returned number of posterior samples; if I am not wrong, you generate those with KDE from the ABC samples (at least this is what happens in the pyABC wrap). In my wrap of ABCpy, I have for now run Rej-ABC for `num_simulations` times, and then simply returned as posterior samples the ones obtained when considering the quantile of distances defined by `quantile`, or the `num_top_samples` ones. I am therefore not returning the specified `num_samples`. Do you suggest using your KDE code in this case as well? I believe it should be easily doable, only did not know what was the precise aim of it. 
- In the pyABC wrap, you choose one of `num_top_samples`, `quantile`, `eps`. I have not yet used the `eps` one, but will add code for that soon.
- In ABCpy, we have the option of generating more than one simulation for parameter value and average the estimate of the distabce; I have not put that option for now, do you think that is useful?
- I've not yet added SASS option (even if that is implemented in ABCpy); also, we do not have postprocessing LRA in ABCpy.

Please let me know what you think of my attempt.

### CODE REVIEW

**Comment by jan-matthis** [comment]
Hi,

first off, this is great, thanks a lot for the PR! 

To answer your questions:
> I am therefore not returning the specified num_samples. Do you suggest using your KDE code in this case as well? I believe it should be easily doable, only did not know what was the precise aim of it.

To compare algorithms we mostly use 2-sample tests, e.g. C2ST. Let's say we run rejection ABC at a simulation budget of 1k simulations and use a quantile of 0.1, so we end up with 100 top samples. Typically, we compute C2STs against 10k reference posterior samples and would like to compute C2ST on a balanced dataset (i.e., containing as many reference as approximate posterior samples). The `num_samples` keyword exists for this reason -- it will typically be `10_000` when we run algorithms, in order to get back the desired amount of samples for calculating metrics.

We could either resample from the population of 100 samples to obtain 10k samples or fit a KDE to obtain more samples -- in the manuscript's appendix, we compare both and found that doing a KDE fit slightly improves performance for some tasks. So it'd be great to have the KDE option for `ABCpy` as well.

(Side note: We also checked computing C2STs using 1k instead of 10k samples -- it did not make much of a difference in terms of overall results -- we opted for 10k to be on the safe side).


> In the pyABC wrap, you choose one of num_top_samples, quantile, eps. I have not yet used the eps one, but will add code for that soon.

We only used quantiles / num_top_samples in the manuscript in the end, but I agree it would be nice to have in case it is easy to do.


> In ABCpy, we have the option of generating more than one simulation for parameter value and average the estimate of the distabce; I have not put that option for now, do you think that is useful?

This sounds very useful to me -- it would allow exploring whether splitting the budget to do multi-simulations per parameter is helpful on a given task.

> I've not yet added SASS option (even if that is implemented in ABCpy); also, we do not have postprocessing LRA in ABCpy.

SASS and LRA were for some extra experiments we report in the appendix, would be nice to have but I think it's totally fine if they are not supported in the first version.

Let me know if more questions come up, or something was unclear in my explanations above.

Best, Jan-Matthis

**Comment by LoryPack** [comment]
Hi Jan-Matthis, 

Thanks for your reply, that was very clear. I'll add the KDE, eps and the multiple simulations per parameter value in the next iteration, and update once that is done. I'll skip the SASS and LRA for now. 

Thanks for your help

**Comment by LoryPack** [comment]
Hello, I have finally had time to work on this (sorry for the long gap from my last update). As suggested, I have done the followin: 

- added the KDE and the resampling to obtain the correct number of samples, following the implementation for the other algorithms
- I have added an option to directly fix the ABC threshold value (`eps`); this is in alternative to either `quantile` or `num_top_samples`. Directly fixing `eps` is however tricky, as it can result in no acceptances among the simulated datasets
- Finally, I have added the option of using multiple simulations for every parameter values in order to better estimate the ABC distance; as you said, this allocates the budges by considering a smaller number of possible parameter values but attempting to estimate better the ABC distance. 

Just as a quick remark, I noticing that some trials raise the following warning when selecting the kernel bandwidth for KDE with cross-validation: 

```
UserWarning: One or more of the test scores are non-finite: [  -11.71028102   -19.40130322   -41.9697046  
 -137.58079662  -1501.8660018             nan            nan            nan            nan            nan]
```

Please let me know if there is anything you'd like to be changed in some way.

File: sbibm/algorithms/abcpy/abcpy_utils.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
+from typing import Callable
```
Comment (jan-matthis) [inline sbibm/algorithms/abcpy/abcpy_utils.py:1]: Could you run `isort` and `black` on the files of the PR? That will take care of sorting imports and keeping the code formatting consistent. Some short info about these two is here: https://github.com/sbi-benchmark/sbibm/blob/main/CONTRIBUTING.md#code-style

File: sbibm/algorithms/abcpy/abcpy_utils.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        self.prior_forward = task.get_prior()
+        self.dim_parameters = task.dim_parameters
+        self.name = task.name if task.name is not None else name
+
+        input_parameters = InputConnector.from_list([])
+        super(ABCpyPrior, self).__init__(input_parameters, self.name + "_prior")
+
+    def forward_simulate(self, abcpy_input_values, num_forward_simulations, rng=np.random.RandomState()):
+        result = np.array(self.prior_forward(num_forward_simulations))
+        # print(num_forward_simulations, result.shape)
```
Comment (jan-matthis) [inline sbibm/algorithms/abcpy/abcpy_utils.py:22]: Can be removed

File: sbibm/algorithms/abcpy/rejection_abc.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    distance_calc = get_distance(distance, statistics)
+
+    # define prior and model from the task simulator:
+    parameters = ABCpyPrior(task)
+    model = ABCpySimulator([parameters], task, max_calls=num_simulations)
+
+    # inference; not sure how to use random seed
+    sampler = RejectionABC(root_models=[model], distances=[distance_calc], backend=backend, seed=None)
+    # print("obs", [observation])
+    # print("obs", [np.array(observation)])
```
Comment (jan-matthis) [inline sbibm/algorithms/abcpy/rejection_abc.py:102]: Can be removed

File: sbibm/algorithms/abcpy/rejection_abc.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    # very large value and then we select the best objects later.
+
+    statistics = Identity(degree=1, cross=True)  # we assume statistics are already computed inside the task
+    distance_calc = get_distance(distance, statistics)
+
+    # define prior and model from the task simulator:
+    parameters = ABCpyPrior(task)
+    model = ABCpySimulator([parameters], task, max_calls=num_simulations)
+
+    # inference; not sure how to use random seed
```
Comment (jan-matthis) [inline sbibm/algorithms/abcpy/rejection_abc.py:99]: Not sure how seeding is handled in `abcpy` but perhaps it's a good idea to introduce an argument that allows passing it?

File: sbibm/algorithms/abcpy/rejection_abc.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    actual_n_simulations = journal_standard_ABC.number_of_simulations[-1] * journal_standard_ABC.configuration[
+        "n_samples_per_param"]
+    # this takes into account the fact that num_simulations can be not divisible by num_simulations_per_param
+    expected_n_simulations = (num_simulations // num_simulations_per_param) * num_simulations_per_param
+    assert actual_n_simulations == expected_n_simulations
+
+    if save_distances:
+        distances = torch.tensor(journal_standard_ABC_reduced.distances[-1])
+        save_tensor_to_csv("/home/lorenzo/Scrivania/OxWaSP/ABC-project/Code/sbibm/distances.csv", distances)
```
Comment (jan-matthis) [inline sbibm/algorithms/abcpy/rejection_abc.py:120]: We should get rid of the path here. The quick solution would be to just do `save_tensor_to_csv("distances.csv", distances)`, so saving it to the current directory, or introducing a new keyword argument

File: try_ABCpy.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import sbibm
```
Comment (jan-matthis) [inline try_ABCpy.py:1]: This file should go out. Either we delete it entirely, or, nicer solution: Parts of it could go into a test case, as parts of `sbibm/tests/algorithms/test_abcpy.py`. The next question would then be what exactly we want to test: E.g. either we have some expectations about how rejection ABC should perform on a subset of tasks or, alternatively, we just check for correctness of returned outputs in terms of shapes

**Review by jan-matthis — CHANGES_REQUESTED** [review]
Hi, thanks a lot, great progress! 

I have done a quick first pass over the PR and left a few comments here and there. Perhaps @janfb can have a look as well?

Regarding the KDE error, could you post a short code snippet to reproduce it (ideally with a fixed random seed)? We'll have a closer look

**Comment by LoryPack** [comment]
Thanks for that, will work on incorporating the updates you suggested. 

Once this is OK for the simple Rejection ABC, I am thinking of wrapping some of the other ABC algorithms which we have in ABCpy, as well as the Synthetic Likelihood ones. 

The issue is however for some ABC algorithms you cannot know a priori how many simulations they will require; I will therefore start by wrapping the ones which have a fixed simulations budget.

File: sbibm/algorithms/abcpy/abcpy_utils.py (hunk 2 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/abcpy_utils.py:22]: Do you mean removing the comment in line 22?

File: sbibm/algorithms/abcpy/rejection_abc.py (hunk 3 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/rejection_abc.py:102]: Sure

File: sbibm/algorithms/abcpy/rejection_abc.py (hunk 4 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/rejection_abc.py:99]: Basically when you instantiate an inference algorithm in ABCpy (which is a class) you pass to it a random seed which is used to initialize a random number generator which is then used everywhere around the inference class. 

I have added a seed option to the `run` function which is used for that. I only did now know whether adding that or not as the other algorithms you implemented do not have it. 

I believe even if you use the abcpy `run` function without explicitly passing the seed but you instead beforehand fix the numpy random seed, reproducibility should be preserved, but the seed can be useful anyway.

File: sbibm/algorithms/abcpy/abcpy_utils.py (hunk 2 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/abcpy_utils.py:22]: I've removed that

File: sbibm/algorithms/abcpy/rejection_abc.py (hunk 5 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/rejection_abc.py:120]: Yes indeed, will remove that path which was for my laptop. Following the implementation in the other algorithms, I will just leave "distances.csv".

File: sbibm/algorithms/abcpy/abcpy_utils.py (hunk 1 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/abcpy_utils.py:1]: Done

File: try_ABCpy.py (hunk 6 above)
Comment (LoryPack) [reply try_ABCpy.py]: I've created a test using the code which was in that file.

File: sbibm/algorithms/abcpy/rejection_abc.py (hunk 4 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/rejection_abc.py]: Actually using the seed in ABCpy does not ensure reproducibility without doing at the start of the script
```
    torch.manual_seed(1)
    np.random.seed(1)
```
that is due to the fact that the simulators do not take a random seed or random number generator as inputs. On the other hand, fixing the above without the ABCpy seed ensures reproducibility. 

Therefore I removed the `seed` parameter from the `run` function for ABCpy, as it was useless. That has therefore the same API as all others.

File: try_ABCpy.py (hunk 6 above)
Comment (LoryPack) [reply try_ABCpy.py]: There I am testing the shape of the output and the actual number of performed simulations using the the possible ways to parametrize the Rejection ABC

File: sbibm/algorithms/abcpy/rejection_abc.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    distance: str = "l2",
+    save_distances: bool = False,
+    sass: bool = False,
+    sass_fraction: float = 0.5,
+    sass_feature_expansion_degree: int = 3,
+    kde_bandwidth: Optional[str] = None,
+) -> Tuple[torch.Tensor, int, Optional[torch.Tensor]]:
+    """Runs REJ-ABC from `ABCpy`
+
+    We do not implement SASS for now, and LRA after sampling.
```
Comment (janfb) [inline sbibm/algorithms/abcpy/rejection_abc.py:41]: this sentence is not clear to me. do you mean that LRA will not be implemented at all and SASS is not yet? Then a suggestion:
"ABCpy does not implement LRA post processing. SASS is supported in ABCpy but not yet implemented here. "

File: sbibm/algorithms/abcpy/rejection_abc.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    parameters = ABCpyPrior(task)
+    model = ABCpySimulator([parameters], task, max_calls=num_simulations)
+
+    # inference
+    sampler = RejectionABC(
+        root_models=[model], distances=[distance_calc], backend=backend
+    )
+    journal_standard_ABC = sampler.sample(
+        [[np.array(observation)]],
+        n_samples=num_simulations // num_simulations_per_param,
```
Comment (janfb) [inline sbibm/algorithms/abcpy/rejection_abc.py:116]: the `n_samples` is a fixed budget, right? there is in principle no way it can be exceeded, and if it was then we would get a `SimulationBudgetExceeded` exception because of the `max_calls` passed above, no?

**Review by janfb — COMMENTED** [review]
Looks great! I added a comment and a question.

File: sbibm/algorithms/abcpy/rejection_abc.py (hunk 7 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/rejection_abc.py]: I fixed that. Sorry, it was very unclear.

File: sbibm/algorithms/abcpy/rejection_abc.py (hunk 8 above)
Comment (LoryPack) [reply sbibm/algorithms/abcpy/rejection_abc.py:116]: Basically in ABCpy Rejection ABC fixes a number of samples with distance from the observation below a given threshold, and simulates from the model until that number of samples are accepted. As a workaround to get a fixed simulation budget, I used an extremely large epsilon so that all simulation are accepted. The results are then post-processed and the ones with smaller distance are accepted. 

I realize it's not the cleanest implementation ever, but it should work here. As in ABCpy we rarely use RejectionABC for practical purposes, we never improved the implementation so that it allows a fixed simulation target.
